source-bigquery-batch/driver.go

```diff
@@ -0,0 +1,698 @@
+const queryDiscoverPrimaryKeys = `
+SELECT kcu.table_schema, kcu.table_name, kcu.column_name, kcu.ordinal_position, col.data_type
+  FROM %[1]s.INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu
+  JOIN %[1]s.INFORMATION_SCHEMA.TABLE_CONSTRAINTS tcs
+    ON  tcs.constraint_catalog = kcu.constraint_catalog
+    AND tcs.constraint_schema = kcu.constraint_schema
+    AND tcs.constraint_name = kcu.constraint_name
+    AND tcs.table_schema = kcu.table_schema
+    AND tcs.table_name = kcu.table_name
+  JOIN %[1]s.INFORMATION_SCHEMA.COLUMNS col
+    ON  col.table_schema = kcu.table_schema
+	AND col.table_name = kcu.table_name
+	AND col.column_name = kcu.column_name
+  WHERE tcs.constraint_type = 'PRIMARY KEY'
+  ORDER BY kcu.table_schema, kcu.table_name, kcu.ordinal_position;
+`
+
+type discoveredPrimaryKey struct {
+	Schema      string
+	Table       string
+	Columns     []string
+	ColumnTypes map[string]*jsonschema.Schema
+}
+
+func discoverPrimaryKeys(ctx context.Context, db *bigquery.Client, dataset string) ([]*discoveredPrimaryKey, error) {
+	var rows, err = db.Query(fmt.Sprintf(queryDiscoverPrimaryKeys, quoteIdentifier(dataset))).Read(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error discovering primary keys: %w", err)
+	}
+
+	var keysByTable = make(map[string]*discoveredPrimaryKey)
+	for {
+		var row []bigquery.Value
+		if err := rows.Next(&row); err == iterator.Done {
+			break
+		} else if err != nil {
+			return nil, fmt.Errorf("error discovering primary keys: %w", err)
+		}
+
+		var tableSchema = row[0].(string)
+		var tableName = row[1].(string)
+		var columnName = row[2].(string)
+		var ordinalPosition = int(row[3].(int64))
+		var dataType = row[4].(string)
+
+		var tableID = tableSchema + "." + tableName
+		var keyInfo = keysByTable[tableID]
+		if keyInfo == nil {
+			keyInfo = &discoveredPrimaryKey{
+				Schema:      tableSchema,
+				Table:       tableName,
+				ColumnTypes: make(map[string]*jsonschema.Schema),
+			}
+			keysByTable[tableID] = keyInfo
+		}
+		keyInfo.Columns = append(keyInfo.Columns, columnName)
+		if jsonSchema, ok := databaseTypeToJSON[dataType]; ok {
+			keyInfo.ColumnTypes[columnName] = jsonSchema
+		} else {
+			// Assume unknown types are strings, because it's almost always a string.
+			// (Or it's an object, but those can't be Flow collection keys anyway)
+			keyInfo.ColumnTypes[columnName] = &jsonschema.Schema{Type: "string"}
+		}
+		if ordinalPosition != len(keyInfo.Columns) {
+			return nil, fmt.Errorf("primary key column %q (of table %q) appears out of order", columnName, tableID)
+		}
+	}
+
+	var keys []*discoveredPrimaryKey
+	for _, key := range keysByTable {
+		keys = append(keys, key)
+	}
+	return keys, nil
+}
+
+var databaseTypeToJSON = map[string]*jsonschema.Schema{
+	"INT64":   {Type: "integer"},
+	"FLOAT64": {Type: "number"},
+}
+
+var catalogNameSanitizerRe = regexp.MustCompile(`(?i)[^a-z0-9\-_.]`)
+
+func recommendedCatalogName(table string) string {
```

Comment:
I know that we're still doing this everywhere, but I don't think connectors actually need to still be doing it since [the runtime does it now](https://github.com/estuary/flow/pull/1083), too. Certainly not something we need to clean up now, but seeing this tickled my memory of that change.

Reply:
Ah, I wasn't aware of that. I'm going to leave this as-is for symmetry with the other connectors for now, but we should definitely start removing that from connectors sometime soon.